clients/mobile/src/userWorkflow/personalInfo/hooks/useAddPersonalInfo.ts

```diff
@@ -0,0 +1,176 @@
+}
+
+export interface PersonalInfo {
+  bloodGroup: string;
+  height: string;
+  weight: string;
+  gender: string;
+  lastDonationDate: Date;
+  dateOfBirth: Date;
+  lastVaccinatedDate: Date;
+  city: string;
+  locations: string[];
+  availableForDonation: string;
+  acceptPolicy: boolean;
+}
+
+interface PersonalInfoErrors extends PersonalInfo {}
+
+const validationRules: Record<PersonalInfoKeys, ValidationRule[]> = {
+  availableForDonation: [validateRequired],
+  city: [validateRequired],
+  locations: [validateRequired],
+  bloodGroup: [validateRequired],
+  lastDonationDate: [validateRequired, validatePastOrTodayDate],
+  height: [validateRequired, validateHeight],
+  weight: [validateRequired, validateWeight],
+  gender: [validateRequired],
+  dateOfBirth: [validateRequired, validateDateOfBirth],
+  lastVaccinatedDate: [validateRequired, validatePastOrTodayDate],
+  acceptPolicy: [validateRequired]
+}
+
+export const useAddPersonalInfo = (): any => {
+  const fetchClient = useFetchClient()
+  const navigation = useNavigation<AddPersonalInfoNavigationProp>()
+  const [personalInfo, setPersonalInfo] = useState<PersonalInfo>({
+    bloodGroup: '',
+    height: '',
+    weight: '',
+    gender: '',
+    lastDonationDate: new Date(new Date().setMonth(new Date().getMonth() - 6)),
+    dateOfBirth: new Date(new Date().setFullYear(new Date().getFullYear() - 18)),
+    lastVaccinatedDate: new Date(new Date().setMonth(new Date().getMonth() - 6)),
+    city: '',
+    locations: [],
+    availableForDonation: 'yes',
+    acceptPolicy: false
+  }
+  )
+  const [errors, setErrors] = useState<PersonalInfoErrors>(
+    initializeState<PersonalInfo>(Object.keys(validationRules) as PersonalInfoKeys[], null)
+  )
+  const [errorMessage, setErrorMessage] = useState<string>('')
+  const [loading, setLoading] = useState<boolean>(false)
+  const [showDatePicker, setShowDatePicker] = useState(false)
+  const [isVisible, setIsVisible] = useState('')
+
+  const handleInputChange = (name: PersonalInfoKeys, value: string): void => {
+    setPersonalInfo(prevState => ({
+      ...prevState,
+      [name]: value
+    }))
+    setShowDatePicker(false)
+    handleInputValidation(name, value)
+  }
+
+  const handleInputValidation = (name: PersonalInfoKeys, value: string | boolean): void => {
+    const errorMsg = validateInput(value as string, validationRules[name])
+    setErrors(prevErrors => ({
+      ...prevErrors,
+      [name]: errorMsg
+    }))
+  }
+
+  const isButtonDisabled = useMemo(() => {
+    return !(
+      Object.values(personalInfo).every(value => value !== '' && !(Array.isArray(value) && value.length === 0)) &&
+      Object.values(errors).every(error => error === null)
+    ) || !personalInfo.acceptPolicy
+  }, [personalInfo, errors])
+
+  async function formatLocations(locations: string[], city: string): Promise<LocationData[]> {
+    const locationService = new LocationService(GOOGLE_MAP_API)
+
+    const formattedLocations = await Promise.all(
```

Comment:
are you sure you want promise.all? as it will cancel if any of the elements error out.

Reply:
Vai, What should we do in this situation? Should we ignore the location that causes the API to fail, notify the user that we couldn't find that location, or should we retry for the failed one?